### PR TITLE
Don't take the lock  unless needed at exit

### PR DESCRIPTION
**Pull request by nouiz**
This lower lock contention.

### CODE REVIEW

**Comment by carriepl** [comment]
Travis is failing

**Comment by nouiz** [comment]
I fixed the tests failure about pep8.

File: theano/gof/cmodule.py [hunk 1]
```diff
@@ -1392,86 +1396,101 @@ def clear_unversioned(self, min_age=None):
...
-            self.clear_old()
-            self.clear_unversioned()
+
+        # Note: no need to take the lock. For unversioned files, we
+        # don't need it as they aren't shared. For old unversioned
+        # files, this happen rarely, so we take the lock only when
+        # this happen.
+
+        # Note: for clear_old(), as this happen unfrequently, we only
+        # take the lock when it happen.
```
Comment (carriepl) [inline theano/gof/cmodule.py]: I think the above information is important enough to be included in the documentation of the methods clear_old() and clear_unversioned().

**Comment by carriepl** [comment]
Other than my tiny comment, I am fine with this PR.

**Comment by carriepl** [comment]
One of the workers on Travis had a timeout. I restarted it so we should have the final diagnostic in ~30 minutes but so far it there doesn't seem to be any issue.

**Comment by carriepl** [comment]
Looks good. Merging.

**Pull request merged**